From 04db4dc2a2ce05063b872eb3cf95c2401c44a9f0 Mon Sep 17 00:00:00 2001
From: Alejandro Vallejo <alejandro.garciavallejo@amd.com>
Date: Wed, 3 Jun 2026 11:37:15 +0200
Subject: xen/gnttab: Fix TOCTOU race in gnttab_set_version()

Move first read of gt->gt_version inside the critical region of the
rwlock, otherwise concurrent gnttab operations (silly as they would be)
may get mutually confused as to the actual current version.

Fixes: c1488502c949("grant-tables: do not fail attempts to...")
Reported-by: Oleksandr Tyshchenko <Oleksandr_Tyshchenko@epam.com>
Signed-off-by: Alejandro Vallejo <alejandro.garciavallejo@amd.com>
Reviewed-by: Jan Beulich <jbeulich@suse.com>
master commit: c0e548d9206b3a281f9d30a2670be543fb383223
master date: 2026-05-22 13:32:44 +0200

diff --git a/xen/common/grant_table.c b/xen/common/grant_table.c
index f9e4bffdb15d..5e0902782163 100644
--- a/xen/common/grant_table.c
+++ b/xen/common/grant_table.c
@@ -3185,11 +3185,12 @@ gnttab_set_version(XEN_GUEST_HANDLE_PARAM(gnttab_set_version_t) uop)
     if ( op.version == 2 && gt->max_version == 1 )
         goto out; /* Behave as before set_version was introduced. */
 
+    grant_write_lock(gt);
+
     res = 0;
     if ( gt->gt_version == op.version )
-        goto out;
+        goto out_unlock;
 
-    grant_write_lock(gt);
     /*
      * Make sure that the grant table isn't currently in use when we
      * change the version number, except for the first 8 entries which
